### PR TITLE
ARTEMIS-1093 Full qualified queue name support

### PR DESCRIPTION
**Pull request by gaohoward**
Broker should support full qualified queue names (FQQN)
as well as bare queue names. This means when clients access
to a queue they have two equivalent ways to do so. One way
is by queue names and the other is by FQQN (i.e. address::qname)
names. Currently only receiving is supported.

### CODE REVIEW

**Comment by gaohoward** [comment]
@clebertsuconic @mtaylor This is a resubmit of FQQN PR, with all tests fixed.

**Comment by gaohoward** [comment]
guys please hold on this commit, I found some garbage files need to remove.

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpSendReceiveTest.java [hunk 1]
```diff
@@ -53,38 +56,61 @@
...
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * Test basic send and receive scenarios using only AMQP sender and receiver links.
  */
+@RunWith(Parameterized.class)
```
Comment (tabish121) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpSendReceiveTest.java]: This seems like an unnecessary change to this test.  The intent here is to test basic AMQP protocol support not to test the FQQN support in the broker which it appears is already done in the ProtonFullQualifiedNameTest.  I'd prefer if we could keep tests targeted and not try and make them test the entirety of the broker features in one monolithic test case as that makes it harder to maintain.  If you changed this one why didn't you change every single AMQP test in the same fashion?

Comment (gaohoward) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/amqp/AmqpSendReceiveTest.java]: @tabish121 ok, I'll remove it to keep it simple.

**Comment by gaohoward** [comment]
OK, I removed the garbage files and reverted AmqpSendReceiveTest.

**Comment by clebertsuconic** [comment]
Please.  While the whole testsuite ?

**Comment by gaohoward** [comment]
There might be a problem with Jenkins machine, the failure relates to 'memory allocation'. Not related to my PR.
@clebertsuconic I ran the whole test suite on my local machine, there are few test failures but none of them related to my PR. They failed both with and without my PR. (I'll investigate those failures further when I got time).

**Comment by clebertsuconic** [comment]
@mtaylor I will let you merge this.. but this PR is +1 from me.. 

nice job @gaohoward

**Comment by clebertsuconic** [comment]
@gaohoward actually, the example managmeent is broken after this. Can you look please?

**Comment by gaohoward** [comment]
@clebertsuconic sure I'll take care of it. Thanks.

**Comment by gaohoward** [comment]
@clebertsuconic @mtaylor turns out I compared a String address with SimpleString, which always return false. Fixed that and added a simple test. 
Jenkins for the moment seems down, I'll kick off a build when it's back.

**Comment by gaohoward** [comment]
Jenkins looking good. :)

File: tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/util/FQQN.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.tests.util;
+
+import org.apache.activemq.artemis.api.core.SimpleString;
+import org.apache.activemq.artemis.utils.CompositeAddress;
+
+public final class FQQN {
+
```
Comment (clebertsuconic) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/util/FQQN.java]: Can you remove this class, and move these two methods towards CompositeAddress

Comment (clebertsuconic) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/util/FQQN.java]: Actually, I already did while merging.

Comment (gaohoward) [inline tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/util/FQQN.java]: @clebertsuconic  Thanks. :)

**Pull request closed without merging**